aten/src/ATen/TensorMeta.h

```diff
@@ -13,6 +13,21 @@ namespace at {
 
 class Tensor;
 
+inline DimVector contiguous_strides(IntArrayRef sizes) {
+  int64_t dims = sizes.size();
+  if (dims == 0) {
+    return {};
+  }
+
+  DimVector strides(dims);
+  strides[dims - 1] = 1;
+  for (auto i = dims - 2; i >= 0; i--) {
+    strides[i] = strides[i + 1] * sizes[i + 1];
+  }
+
+  return strides;
+}
```

Comment:
I swear we should have another implementation of this lying around somewhere...

Reply:
In fact, not only we do, but this one happens to be incorrect! Strides should not be zero even when the sizes are zero.

I implemented this one, which also gives fortran-contiguous strides as a helper for linalg. It is based on how TI generates the strides.

https://github.com/pytorch/pytorch/blob/b17b2b1cc7b017c3daaeff8cc7ec0f514d42ec37/aten/src/ATen/native/LinearAlgebraUtils.h#L36-L66

Reply:
fwiw, I think we should put this implementation (or an equivalent one that works just for contiguous tensors) into a more visible header.

Reply:
> I swear we should have another implementation of this lying around somewhere...

As far as I know, there are these 2:

- [`jit_type.h`](https://github.com/pytorch/pytorch/blob/0708630d9f30210b0a51d13558ad173da630ef95/aten/src/ATen/core/jit_type.h#L789-L798)
- [`LinearAlgebraUtils.h`](https://github.com/pytorch/pytorch/blob/0708630d9f30210b0a51d13558ad173da630ef95/aten/src/ATen/native/LinearAlgebraUtils.h#L36-L66)

> fwiw, I think we should put this implementation (or an equivalent one that works just for contiguous tensors) into a more visible header.

Agreed. That said, I'm not sure where to put it myself. This was the main reason I created one here.
Anyway, I will try to find a better place to leave it.

Reply:
This is general enough that it could be somewhere in c10, I'd say.